File: EquiDuty/EquiDuty/Features/Facilities/ViewModels/FacilityDetailViewModel.swift
//
//  FacilityDetailViewModel.swift
//  EquiDuty
//
//  ViewModel for facility detail view with day schedule
//

import Foundation

@MainActor
@Observable
final class FacilityDetailViewModel {
    // MARK: - State

    var facility: Facility?
    var reservations: [FacilityReservation] = []
    var availableSlots: AvailableSlotsResponse?
    var selectedDate: Date = .now
    var isLoading = false
    var errorMessage: String?

    // MARK: - Dependencies

    private let facilityService = FacilityService.shared
    private let reservationService = FacilityReservationService.shared
    private let listenerService = ReservationListenerService()
    private let dateFormatter: DateFormatter
    let facilityId: String

    init(facilityId: String) {
        self.facilityId = facilityId
        dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"
        dateFormatter.locale = Locale(identifier: "en_US_POSIX")
    }

    // MARK: - Computed

    var facilityName: String {
        facility?.name ?? ""
    }

    var facilityType: String {
        facility?.type ?? ""
    }

    // MARK: - Real-time listener

    /// Start the Firestore listener for the facility's stable.
    /// Call from .onAppear or .task { }.
    func startRealtimeUpdates(stableId: String) {
        listenerService.startListening(stableId: stableId)
    }

    /// Stop the Firestore listener.
    /// Call from .onDisappear.
    func stopRealtimeUpdates() {
        listenerService.stopListening()
    }

    /// Reservations from the real-time listener, filtered to current facility + date.
    /// Falls back to API-fetched reservations if listener has no data yet.
    var liveReservations: [FacilityReservation] {
        let source = listenerService.reservations.isEmpty ? reservations : listenerService.reservations
        let calendar = Calendar.current
        return source.filter { reservation in
            reservation.facilityId == facilityId &&
            calendar.isDate(reservation.startTime, inSameDayAs: selectedDate)
        }
    }

    // MARK: - Actions

    func loadData() async {
        isLoading = true
        errorMessage = nil

        do {
            async let facilityResult = facilityService.getFacility(id: facilityId)
            async let slotsResult = facilityService.getAvailableSlots(facilityId: facilityId, date: selectedDate)
            async let reservationsResult = loadReservationsForDate()

            facility = try await facilityResult
            availableSlots = try await slotsResult
            reservations = try await reservationsResult
            isLoading = false

            // Start listener once we know the stableId
            if let stableId = facility?.stableId {
                startRealtimeUpdates(stableId: stableId)
            }
        } catch {
            errorMessage = error.localizedDescription
            isLoading = false
        }
    }

    func changeDate(_ date: Date) async {
        selectedDate = date
        do {
            async let slotsResult = facilityService.getAvailableSlots(facilityId: facilityId, date: date)
            async let reservationsResult = loadReservationsForDate()

            availableSlots = try await slotsResult
            reservations = try await reservationsResult
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func goToPreviousDay() async {
        let newDate = Calendar.current.date(byAdding: .day, value: -1, to: selectedDate) ?? selectedDate
        await changeDate(newDate)
    }

    func goToNextDay() async {
        let newDate = Calendar.current.date(byAdding: .day, value: 1, to: selectedDate) ?? selectedDate
        await changeDate(newDate)
    }

    // MARK: - Private

    private func loadReservationsForDate() async throws -> [FacilityReservation] {
        try await reservationService.getReservations(
            facilityId: facilityId,
            startDate: selectedDate,
            endDate: selectedDate
        )
    }
}

File: EquiDuty-Android/app/src/main/java/com/equiduty/ui/facilities/FacilityDetailViewModel.kt
package com.equiduty.ui.facilities

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.equiduty.data.repository.FacilityRepository
import com.equiduty.data.repository.FacilityReservationRepository
import com.equiduty.data.repository.ReservationListenerRepository
import com.equiduty.domain.model.Facility
import com.equiduty.domain.model.FacilityReservation
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.launch
import timber.log.Timber
import java.time.LocalDate
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import javax.inject.Inject

@HiltViewModel
class FacilityDetailViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val facilityRepository: FacilityRepository,
    private val reservationRepository: FacilityReservationRepository,
    private val listenerRepository: ReservationListenerRepository
) : ViewModel() {

    private val facilityId: String = savedStateHandle["facilityId"] ?: ""

    private val _facility = MutableStateFlow<Facility?>(null)
    val facility: StateFlow<Facility?> = _facility.asStateFlow()

    private val _reservations = MutableStateFlow<List<FacilityReservation>>(emptyList())
    val reservations: StateFlow<List<FacilityReservation>> = _reservations.asStateFlow()

    private val _selectedDate = MutableStateFlow(LocalDate.now())
    val selectedDate: StateFlow<LocalDate> = _selectedDate.asStateFlow()

    private val _isLoading = MutableStateFlow(true)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    /**
     * Live reservations from Firestore listener, filtered to current facility + date.
     * Falls back to API-fetched reservations if listener has no data yet.
     */
    val liveReservations: StateFlow<List<FacilityReservation>> = combine(
        listenerRepository.liveReservations,
        _reservations,
        _selectedDate
    ) { live, apiFetched, date ->
        val source = live.ifEmpty { apiFetched }
        val dateStr = date.format(DateTimeFormatter.ISO_LOCAL_DATE)
        source.filter { reservation ->
            reservation.facilityId == facilityId &&
            reservation.startTime.startsWith(dateStr)
        }
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    init {
        loadFacility()
        loadReservationsForDate()
    }

    private fun loadFacility() {
        viewModelScope.launch {
            try {
                val facility = facilityRepository.getFacility(facilityId)
                _facility.value = facility

                // Start real-time listener once we know the stableId
                facility?.stableId?.let { stableId ->
                    listenerRepository.startListening(stableId)
                }
            } catch (e: Exception) {
                Timber.e(e, "Failed to load facility")
            } finally {
                _isLoading.value = false
            }
        }
    }

    fun loadReservationsForDate(date: LocalDate = _selectedDate.value) {
        _selectedDate.value = date
        viewModelScope.launch {
            try {
                val dateStr = date.format(DateTimeFormatter.ISO_LOCAL_DATE)
                reservationRepository.fetchReservations(
                    facilityId = facilityId,
                    startDate = dateStr,
                    endDate = dateStr
                )
                _reservations.value = reservationRepository.reservations.value
            } catch (e: Exception) {
                Timber.e(e, "Failed to load reservations for date")
            }
        }
    }

    fun navigateDate(days: Int) {
        val newDate = _selectedDate.value.plusDays(days.toLong())
        loadReservationsForDate(newDate)
    }

    override fun onCleared() {
        super.onCleared()
        listenerRepository.stopListening()
    }
}
